File: AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        add = 0
        head = ListNode(0)
        l3 = head
        while l1 or l2:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0        
            
            sum = num1 + num2 + add
            curval = sum %10
            add = sum/10
            l3.next = ListNode(curval)
            l3 = l3.next
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        if add:
            l3.next = ListNode(add)
            l3 = l3.next
        #l3.val = add
        return head.next

File: addTwoNumbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var h1 = l1
        var h2 = l2
        var carry = 0
        var result = ListNode(0)
        var resultH = result
        while (h1 != null && h2 != null){
            val tmp = h1.`val` + h2.`val` + carry
            val newVal = tmp % 10
            carry = tmp / 10
            val newNode = ListNode(newVal)
            resultH.next = newNode
            h1 = h1.next
            h2 = h2.next
            resultH = resultH.next
        }

        while (h1 != null) {
            val tmp = h1.`val` + carry
            val newVal = tmp % 10
            carry = tmp / 10
            val newNode = ListNode(newVal)
            resultH.next = newNode
            h1 = h1.next
            resultH = resultH.next
        }
        while (h2 != null) {
            val tmp = h2.`val` + carry
            val newVal = tmp % 10
            carry = tmp / 10
            val newNode = ListNode(newVal)
            resultH.next = newNode
            h2 = h2.next
            resultH = resultH.next
        }
        if (carry != 0) {
            val newNode = ListNode(carry)
            resultH.next = newNode
        }
        return result.next
    }
}
